vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/AndroidDeviceTestRunner.kt

```diff
@@ -10,46 +10,40 @@ import com.malinskiy.marathon.android.AndroidConfiguration
 import com.malinskiy.marathon.android.AndroidDevice
 import com.malinskiy.marathon.android.ApkParser
 import com.malinskiy.marathon.android.InstrumentationInfo
-import com.malinskiy.marathon.android.executor.listeners.CompositeTestRunListener
-import com.malinskiy.marathon.android.executor.listeners.DebugTestRunListener
-import com.malinskiy.marathon.android.executor.listeners.LogCatListener
-import com.malinskiy.marathon.android.executor.listeners.NoOpTestRunListener
-import com.malinskiy.marathon.android.executor.listeners.ProgressTestRunListener
-import com.malinskiy.marathon.android.executor.listeners.TestRunResultsListener
-import com.malinskiy.marathon.android.executor.listeners.screenshot.ScreenCapturerTestRunListener
-import com.malinskiy.marathon.android.executor.listeners.video.ScreenRecorderTestRunListener
 import com.malinskiy.marathon.android.safeClearPackage
-import com.malinskiy.marathon.device.DeviceFeature
-import com.malinskiy.marathon.device.DevicePoolId
 import com.malinskiy.marathon.exceptions.DeviceLostException
 import com.malinskiy.marathon.exceptions.TestBatchExecutionException
 import com.malinskiy.marathon.execution.Configuration
-import com.malinskiy.marathon.execution.TestBatchResults
-import com.malinskiy.marathon.execution.progress.ProgressReporter
-import com.malinskiy.marathon.io.FileManager
 import com.malinskiy.marathon.log.MarathonLogging
-import com.malinskiy.marathon.report.attachment.AttachmentProvider
-import com.malinskiy.marathon.report.logs.LogWriter
 import com.malinskiy.marathon.test.MetaProperty
 import com.malinskiy.marathon.test.Test
 import com.malinskiy.marathon.test.TestBatch
 import com.malinskiy.marathon.test.toTestName
-import kotlinx.coroutines.CompletableDeferred
 import java.io.IOException
 import java.util.concurrent.TimeUnit
-
-val JUNIT_IGNORE_META_PROPERY = MetaProperty("org.junit.Ignore")
+import kotlin.math.max
+import kotlin.math.min
 
 class AndroidDeviceTestRunner(private val device: AndroidDevice) {
 
+    companion object {
+        private val JUNIT_IGNORE_META_PROPERTY = MetaProperty("org.junit.Ignore")
+        private const val MAX_TEST_DURATION_LEEWAY = 1.2 // +20% to expected test duration
+        private const val BATCH_DURATION_LEEWAY = 1.1 // +10% to expected batch duration
+    }
+
     private val logger = MarathonLogging.logger("AndroidDeviceTestRunner")
 
-    fun execute(configuration: Configuration,
-                rawTestBatch: TestBatch,
-                listener: ITestRunListener) {
+    fun execute(
+        configuration: Configuration,
+        rawTestBatch: TestBatch,
+        listener: ITestRunListener
+    ) {
+        val ignoredTests = rawTestBatch.tests.filter { it.metaProperties.contains(JUNIT_IGNORE_META_PROPERTY) }
 
-        val ignoredTests = rawTestBatch.tests.filter { it.metaProperties.contains(JUNIT_IGNORE_META_PROPERY) }
-        val testBatch = TestBatch(rawTestBatch.tests - ignoredTests)
```

Comment:
Are you sure this is an appropriate place to remove ignored tests from execution at all? Shouldn't they be handled on actor level?

Reply:
No, I also think this is not the right place for ignoring such tests. But that is out of scope for this PR and it was not changed in my PR.